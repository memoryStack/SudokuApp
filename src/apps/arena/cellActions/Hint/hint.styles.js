import { StyleSheet } from 'react-native'

export const getStyles = ({}, theme) => StyleSheet.create({

})
