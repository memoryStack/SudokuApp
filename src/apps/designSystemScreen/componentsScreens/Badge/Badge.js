import React, { memo } from 'react'

import { View } from 'react-native'

import Text from '@ui/atoms/Text'
import Badge from '@ui/atoms/Badge'

import { Page } from 'src/apps/components/Page'

import { styles } from './badge.styles'

const BadgeDemo = () => (
    <Page style={styles.container}>
        <View>
            <View style={{ height: 32, width: 32, backgroundColor: 'pink' }}>
                <Text>Icon</Text>
            </View>
            <Badge label="99999+" />
        </View>
    </Page>
)

export default memo(BadgeDemo)
