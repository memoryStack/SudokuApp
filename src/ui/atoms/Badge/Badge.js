import React from 'react'

import { ViewPropTypes } from 'react-native'

import PropTypes from 'prop-types'

import { useStyles } from '@utils/customHooks/useStyles'

import Text from '../Text'

import { BADGE_TEST_ID, BADGE_TYPE } from './badge.constants'
import { getStyles } from './badge.styles'

/*
    How to use this properly
        <View style={{  // this view will bind the icon and the badge together
            marginTop: 40,
            borderWidth: 2,
            borderColor: 'black',
        }}>
            // this view will decide the layout sizing, it will be an Icon/Image
            // or some other view which will have the badge over it
            <View style={{ height: 32,width: 32 }} />

            // this is absolutely positioned anyway
            <Badge label="99999+" />
        </View>

        Note: the view which is binding Icon/Image should be not limited for space
            else the Badge will shrink and won't be able to expand properly
        TODO: this demands a research if an absolutely positioned element can be
        intrinsicly(without height/width set) wider than it's containing block.
*/

const Badge = ({
    type,
    label,
    styles: stylesProp,
}) => {
    const styles = useStyles(getStyles, { type })

    const renderLabel = () => (type === BADGE_TYPE.SMALL ? '' : label)

    return (
        <Text
            testID={BADGE_TEST_ID}
            style={[styles.label, styles.container, stylesProp]}
        >
            {renderLabel()}
        </Text>
    )
}

export default React.memo(Badge)

Badge.propTypes = {
    type: PropTypes.string,
    label: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.string,
    ]),
    styles: ViewPropTypes.style,
}

Badge.defaultProps = {
    type: BADGE_TYPE.LARGE,
    label: '',
    styles: {},
}
