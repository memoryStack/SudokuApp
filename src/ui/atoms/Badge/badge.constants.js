export const BADGE_TYPE = {
    SMALL: 'sm',
    LARGE: 'lg',
}
