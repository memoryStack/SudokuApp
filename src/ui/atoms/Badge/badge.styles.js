import { StyleSheet } from 'react-native'

import _get from '@lodash/get'

import { BADGE_TYPE } from './badge.constants'

export const getStyles = ({ type }, theme) => {
    const size = _get(theme, ['badge', type, 'container', 'layout', 'size'])

    const sizeProps = {
        height: size,
        ...type === BADGE_TYPE.LARGE && { minWidth: size },
        ...type === BADGE_TYPE.SMALL && { width: size },
    }

    // NOTE: after assuming that all the icons/images with badges on them
    //      will be of 32*32 need to standardize the icon sizes
    const placementProps = {
        top: type === BADGE_TYPE.SMALL ? 0 : -4,
        ...type === BADGE_TYPE.SMALL && { right: 0 },
        ...type === BADGE_TYPE.LARGE && { left: 16 },
    }

    return StyleSheet.create({
        container: {
            display: 'flex',
            alignItems: 'center',
            justifyContent: 'center',
            position: 'absolute',
            ...placementProps,
            backgroundColor: _get(theme, ['badge', type, 'container', 'color']),
            ..._get(theme, ['badge', type, 'container', 'layout', 'shape']),
            padding: _get(theme, ['badge', type, 'container', 'layout', 'padding']),
            ...sizeProps,
        },
        label: {
            ...type === BADGE_TYPE.LARGE && {
                color: _get(theme, ['badge', type, 'label', 'color']),
                ..._get(theme, ['badge', type, 'label', 'font']),
            },
        },
    })
}
