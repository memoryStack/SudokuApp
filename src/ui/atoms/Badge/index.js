export { default } from './Badge'
export * from './badge.constants'
