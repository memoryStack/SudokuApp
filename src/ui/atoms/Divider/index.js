export { default } from './Divider'
export * from './divider.constants'
