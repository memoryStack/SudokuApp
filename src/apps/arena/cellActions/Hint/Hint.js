import React from 'react'

import PropTypes from 'prop-types'

import _noop from '@lodash/noop'

import { HintIcon } from '@resources/svgIcons/hint'

import Badge from '@ui/atoms/Badge'
import Text from '@ui/atoms/Text'

import { Touchable } from '../../../components/Touchable'

import { styles, INACTIVE_ICON_FILL } from '../cellActions.styles'

// TODO: i should make it as a part of settings so that users can change it according to their confidence level
// and also we can make the hints numbers vary according to the difficulty level. user can customize that as per their
// comfort and confidence level

const Hint = ({
    iconBoxSize, hints, onClick, disabled, ...rest
}) => {
    const renderHintsCount = () => {
        if (disabled) return null
        return <Badge label={hints} />
    }

    const renderIcon = () => (
        <>
            <HintIcon iconBoxSize={iconBoxSize} fill={INACTIVE_ICON_FILL} />
            {renderHintsCount()}
        </>
    )

    const noAvailableHints = hints === 0

    return (
        <Touchable style={styles.actionContainer} onPress={onClick} disabled={disabled || noAvailableHints} {...rest}>
            {renderIcon()}
            <Text style={styles.actionText}>Hint</Text>
        </Touchable>
    )
}

export default React.memo(Hint)

Hint.propTypes = {
    iconBoxSize: PropTypes.number,
    onClick: PropTypes.func,
    hints: PropTypes.number,
    disabled: PropTypes.bool,
}

Hint.defaultProps = {
    iconBoxSize: 40,
    onClick: _noop,
    hints: 0,
    disabled: false,
}
