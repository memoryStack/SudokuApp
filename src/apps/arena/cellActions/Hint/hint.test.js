import * as React from 'react'

import {
    fireEvent, render, screen,
} from '@utils/testing/testingLibrary'

import { BADGE_TEST_ID } from '@ui/atoms/Badge'
import Hint from './Hint'

test('snapshot for enabled Hint', () => {
    const tree = render(<Hint iconBoxSize={50} hints={3} />).toJSON()
    expect(tree).toMatchSnapshot()
})

test('snapshot for disabled Hint', () => {
    const tree = render(<Hint iconBoxSize={50} hints={3} />).toJSON()
    expect(tree).toMatchSnapshot()
})

describe('Hint', () => {
    test.skip('will show remaining hints count in a Badge', () => {
        render(<Hint iconBoxSize={50} hints={3} />)

        expect(screen.getByTestId(BADGE_TEST_ID)).toHaveTextContent(3)
    })

    test('will call handler on click', () => {
        const onClick = jest.fn()
        render(<Hint iconBoxSize={50} onClick={onClick} hints={3} />)

        fireEvent.press(screen.getByText('Hint'))

        expect(onClick).toHaveBeenCalledTimes(1)
    })

    test('if available hints are 0, will not call handler on click', () => {
        const onClick = jest.fn()
        render(<Hint iconBoxSize={50} onClick={onClick} hints={0} />)

        fireEvent.press(screen.getByText('Hint'))

        expect(onClick).toHaveBeenCalledTimes(0)
    })

    test('if disabled, will not call handler on click', () => {
        const onClick = jest.fn()
        render(<Hint iconBoxSize={50} onClick={onClick} disabled hints={3} />)

        fireEvent.press(screen.getByText('Hint'))

        expect(onClick).not.toHaveBeenCalled()
    })

    test.skip('if disabled, hints badge will not be displayed', () => {
        render(<Hint iconBoxSize={50} hints={3} disabled />)

        expect(screen.queryByTestId(BADGE_TEST_ID)).not.toBeOnTheScreen()
    })
})
