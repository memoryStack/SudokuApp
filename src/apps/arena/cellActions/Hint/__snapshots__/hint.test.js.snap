// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot for disabled Hint 1`] = `
<View
  accessibilityState={
    Object {
      "disabled": false,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  hitSlop={null}
  nativeID="animatedComponent"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "justifyContent": "center",
      "opacity": 1,
      "width": 60,
    }
  }
>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight={50}
    bbWidth={50}
    focusable={false}
    height={50}
    meetOrSlice={0}
    minX={0}
    minY={0}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "flex": 0,
          "height": 50,
          "width": 50,
        },
      ]
    }
    vbHeight={512}
    vbWidth={352}
    width={50}
  >
    <RNSVGGroup>
      <RNSVGPath
        d="M96.06 454.35c.01 6.29 1.87 12.45 5.36 17.69l17.09 25.69a31.99 31.99 0 0 0 26.64 14.28h61.71a31.99 31.99 0 0 0 26.64-14.28l17.09-25.69a31.989 31.989 0 0 0 5.36-17.69l.04-38.35H96.01l.05 38.35zM0 176c0 44.37 16.45 84.85 43.56 115.78 16.52 18.85 42.36 58.23 52.21 91.45.04.26.07.52.11.78h160.24c.04-.26.07-.51.11-.78 9.85-33.22 35.69-72.6 52.21-91.45C335.55 260.85 352 220.37 352 176 352 78.61 272.91-.3 175.45 0 73.44.31 0 82.97 0 176zm176-80c-44.11 0-80 35.89-80 80 0 8.84-7.16 16-16 16s-16-7.16-16-16c0-61.76 50.24-112 112-112 8.84 0 16 7.16 16 16s-7.16 16-16 16z"
        fill={4286545791}
        propList={
          Array [
            "fill",
          ]
        }
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <Text
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#CC5D00",
        "borderRadius": 1000,
        "color": "#FFFFFF",
        "display": "flex",
        "fontFamily": "OpenSans-Medium",
        "fontSize": 11,
        "fontWeight": "500",
        "height": 16,
        "includeFontPadding": false,
        "justifyContent": "center",
        "left": 30,
        "letterSpacing": 0.5,
        "lineHeight": 16,
        "minWidth": 16,
        "padding": 0,
        "position": "absolute",
        "textAlign": "center",
        "textAlignVertical": "center",
        "top": -4,
      }
    }
    testID="badge_test_id"
  >
    3
  </Text>
  <Text
    style={
      Object {
        "color": "rgb(127, 127, 127)",
        "fontFamily": "OpenSans",
        "fontSize": 16,
        "marginTop": 8,
      }
    }
  >
    Hint
  </Text>
</View>
`;

exports[`snapshot for enabled Hint 1`] = `
<View
  accessibilityState={
    Object {
      "disabled": false,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  hitSlop={null}
  nativeID="animatedComponent"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "justifyContent": "center",
      "opacity": 1,
      "width": 60,
    }
  }
>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight={50}
    bbWidth={50}
    focusable={false}
    height={50}
    meetOrSlice={0}
    minX={0}
    minY={0}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "flex": 0,
          "height": 50,
          "width": 50,
        },
      ]
    }
    vbHeight={512}
    vbWidth={352}
    width={50}
  >
    <RNSVGGroup>
      <RNSVGPath
        d="M96.06 454.35c.01 6.29 1.87 12.45 5.36 17.69l17.09 25.69a31.99 31.99 0 0 0 26.64 14.28h61.71a31.99 31.99 0 0 0 26.64-14.28l17.09-25.69a31.989 31.989 0 0 0 5.36-17.69l.04-38.35H96.01l.05 38.35zM0 176c0 44.37 16.45 84.85 43.56 115.78 16.52 18.85 42.36 58.23 52.21 91.45.04.26.07.52.11.78h160.24c.04-.26.07-.51.11-.78 9.85-33.22 35.69-72.6 52.21-91.45C335.55 260.85 352 220.37 352 176 352 78.61 272.91-.3 175.45 0 73.44.31 0 82.97 0 176zm176-80c-44.11 0-80 35.89-80 80 0 8.84-7.16 16-16 16s-16-7.16-16-16c0-61.76 50.24-112 112-112 8.84 0 16 7.16 16 16s-7.16 16-16 16z"
        fill={4286545791}
        propList={
          Array [
            "fill",
          ]
        }
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <Text
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#CC5D00",
        "borderRadius": 1000,
        "color": "#FFFFFF",
        "display": "flex",
        "fontFamily": "OpenSans-Medium",
        "fontSize": 11,
        "fontWeight": "500",
        "height": 16,
        "includeFontPadding": false,
        "justifyContent": "center",
        "left": 30,
        "letterSpacing": 0.5,
        "lineHeight": 16,
        "minWidth": 16,
        "padding": 0,
        "position": "absolute",
        "textAlign": "center",
        "textAlignVertical": "center",
        "top": -4,
      }
    }
    testID="badge_test_id"
  >
    3
  </Text>
  <Text
    style={
      Object {
        "color": "rgb(127, 127, 127)",
        "fontFamily": "OpenSans",
        "fontSize": 16,
        "marginTop": 8,
      }
    }
  >
    Hint
  </Text>
</View>
`;
